python/build/lib/pcc/dataframe.py

```diff
@@ -1,149 +1,835 @@
+        universe = []
+        param_list = []
+        for tp in pcctype.__ENTANGLED_TYPES__:
+            universe.append(relevant_objs[tp])
+
+        # Creating random rule for parameters, all Collections first, then singleton objects
+        # Why? Random, couldnt think of a better way
+
+        if hasCollection:
+            for tp in pcctype.__parameter_types__[ParameterMode.Collection]:
+                param_list.append(param_map[tp])
+        if hasSingleton:
+            param_list.extend(params)
+
+        try:
+            pcc_objects = create(pcctype, *universe, params = param_list)
+        except TypeError, e:
+            print ("Exception in __make_pcc: " + e.message)
+            return []
+        return pcc_objects
+
+    def __construct_pccs(self, pcctype, pccs, universe, param):
+        if pcctype in pccs:
+            return pccs[pcctype]
+        params, paramtypes = [], []
+        hasSingleton = False
+        hasCollection = False
+        if hasattr(pcctype, "__parameter_types__"):
+            if ParameterMode.Singleton in pcctype.__parameter_types__:
+                if param == None or len(pcctype.__parameter_types__[ParameterMode.Singleton]) != len(param):
+                    raise TypeError("Connection construct parameter type %s without right parameters" % pcctype.__realname__)
+                params = param
+                hasSingleton = True
+            if ParameterMode.Collection in pcctype.__parameter_types__:
+                paramtypes = list(pcctype.__parameter_types__[ParameterMode.Collection])
+                hasCollection = True
+        
+        dependent_types = list(pcctype.__ENTANGLED_TYPES__)
+        dependent_pccs = [tp for tp in (dependent_types + paramtypes) if tp.__realname__ in self.impure or tp.__realname__ not in universe]
+        if len([tp for tp in dependent_pccs if tp.__PCC_BASE_TYPE__]) > 0:
+            pccs[pcctype] = []
+            return pccs[pcctype]
+        to_be_resolved = [tp for tp in dependent_pccs if tp not in pccs]
+        
+        for tp in to_be_resolved:
+            self.__construct_pccs(tp, pccs, universe, None)
+        param_collection_objs = dict([(tp,
+               universe[tp.__realname__].values()
+                 if tp not in pccs else
+               pccs[tp]) for tp in paramtypes])
+
+        relevant_objs = dict([(tp,
+               universe[tp.__realname__].values()
+                 if tp not in pccs else
+               pccs[tp]) for tp in dependent_types])
+
+        pccs[pcctype] = self.__make_pcc(pcctype, relevant_objs, param_collection_objs, params, hasSingleton, hasCollection)
+
+    def __calculate_pcc(self, pcctypes, universe, params, force = False):
+        if not (self.calculate_pcc or force):
+           return {} 
+        pccs = {}
+        for pcctype in pcctypes:
+            self.__construct_pccs(pcctype, pccs, universe, params)
+        return pccs
+
+    def __process_record(self, record, nsp):
```

Comment:
Im not sure what the return of this method is supposed to be. I see several something or nsp as a return value, not sure what that will return